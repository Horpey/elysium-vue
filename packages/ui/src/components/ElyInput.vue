<script lang="ts" setup>
import type { FunctionalComponent, InputHTMLAttributes } from 'vue'
import ElyLoader from '../components/ElyLoader.vue'
import type { ElyInputSizeKey } from '~/types/ElyInput'

interface Props extends /* @vue-ignore */ InputHTMLAttributes {
  type?: 'text' | 'password' | 'email' | 'number' | 'tel' | 'url'
  name?: string
  modelValue?: string | number
  required?: boolean
  placeholder?: string
  disabled?: boolean
  loading?: boolean
  inputClass?: string
  inputId?: string
  size?: ElyInputSizeKey
  icon?: FunctionalComponent
  onInput?: (event: Event) => void
  onBlur?: (event: Event) => void
  onChange?: (event: Event) => void
}

const props = withDefaults(defineProps<Props>(), {
  type: 'text',
  name: undefined,
  modelValue: undefined,
  required: false,
  placeholder: undefined,
  disabled: false,
  loading: false,
  inputClass: undefined,
  class: undefined,
  inputId: undefined,
  size: 'MEDIUM',
  icon: undefined,
  onInput: () => { },
  onBlur: () => { },
  onChange: () => { },
})

const emit = defineEmits(['update:modelValue'])

const input = ref<HTMLInputElement | null>(null)

const classes = computed(() => {
  const baseClass = 'border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900'

  let classString = baseClass
  let sizeClass = ''

  switch (props.size) {
    case 'SMALL':
      sizeClass = 'text-xs px-2 py-1'
      break
    case 'MEDIUM':
      sizeClass = 'text-sm px-3 py-2'
      break
    case 'LARGE':
      sizeClass = 'text-base px-4 py-3'
      break
  }

  if (props.icon) {
    switch (props.size) {
      case 'SMALL':
        sizeClass = 'text-xs px-2 py-1 pl-6'
        break
      case 'MEDIUM':
        sizeClass = 'text-sm px-3 py-2 pl-8'
        break
      case 'LARGE':
        sizeClass = 'text-base px-4 py-3 pl-10'
        break
    }
  }

  if (props.class)
    classString += ` ${props.class}`

  if (props.inputClass)
    classString += ` ${props.inputClass}`

  if (sizeClass)
    classString += ` ${sizeClass}`

  return classString
})

function onInput(event: Event) {
  props.onInput(event)
  emit('update:modelValue', (event.target as HTMLInputElement).value)
}

function onBlur(event: Event) {
  props.onBlur(event)
}

function onChange(event: Event) {
  props.onChange(event)
}
</script>

<template>
  <span class="relative">
    <input
      :id="inputId"
      ref="input"
      :class="classes"
      :name="name"
      :value="modelValue"
      :type="type"
      :required="required"
      :placeholder="placeholder"
      :disabled="disabled || loading"
      v-bind="$attrs"
      @input="onInput"
      @blur="onBlur"
      @change="onChange"
    >

    <span
      class="pointer-events-none absolute inset-y-0 start-0 flex items-center"
      :class="{ 'pl-2': size === 'SMALL', 'pl-3': size === 'MEDIUM', 'pl-4': size === 'LARGE' }"
    >
      <component
        :is="icon"
        v-if="icon"
        class="text-koromiko-600"
        :class="{
          'left-2 h-3 w-3': size === 'SMALL',
          'left-3 h-4 w-4': size === 'MEDIUM',
          'left-4 h-5 w-5': size === 'LARGE',
        }"
      />
    </span>

    <ElyLoader
      v-if="loading" size="SMALL" class="absolute top-1" :class="{
        'right-2': size === 'SMALL',
        'right-3': size === 'MEDIUM',
        'right-4': size === 'LARGE',
      }"
    />
    <slot />
  </span>
</template>
