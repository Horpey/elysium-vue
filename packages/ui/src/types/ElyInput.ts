export const ELY_INPUT_SIZE = {
  LARGE: 'large',
  MEDIUM: 'medium',
  SMALL: 'small',
} as const

export type ElyInputSizeKey = keyof typeof ELY_INPUT_SIZE
