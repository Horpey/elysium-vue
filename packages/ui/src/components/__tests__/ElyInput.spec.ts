import { mount } from '@vue/test-utils'
import { describe, expect, it, vi } from 'vitest'
import ElyInput from '../ElyInput.vue'

describe('elyInput', () => {
  it('renders with default props', async () => {
    const wrapper = mount(ElyInput)

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert default values
    expect(wrapper.find('input').exists()).toBe(true)
    expect(wrapper.find('input').attributes('type')).toBe('text')
    expect(wrapper.find('input').classes()).toContain('border-koromiko-600')

    // Assert slot content is not provided
    expect(wrapper.text()).toBe('')

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders with custom props and triggers input event', async () => {
    const wrapper = mount(ElyInput, {
      props: {
        type: 'email',
        name: 'email',
        modelValue: 'test@example.com',
        required: true,
        placeholder: 'Enter your email',
        disabled: false,
        loading: false,
        inputClass: 'custom-input-class',
        size: 'MEDIUM',
        onInput: vi.fn(),
        onBlur: vi.fn(),
        onChange: vi.fn(),
      },
    })

    // Assert that the component is rendered
    expect(wrapper.exists()).toBe(true)

    // Assert custom values
    expect(wrapper.find('input').attributes('type')).toBe('email')
    expect(wrapper.find('input').attributes('name')).toBe('email')
    expect(wrapper.find('input').attributes('placeholder')).toBe('Enter your email')
    expect(wrapper.find('input').classes()).toContain('custom-input-class')

    // Trigger input event
    await wrapper.find('input').setValue('new-email@example.com')

    // Assert that the onInput function was called
    expect(wrapper.props().onInput).toHaveBeenCalled()

    // Assert snapshot
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('icon', () => {
    it.each(['SMALL', 'MEDIUM', 'LARGE'])('size %s', (size) => {
      const wrapper = mount(ElyInput, {
        props: {
          type: 'email',
          name: 'email',
          modelValue: 'test@example.com',
          placeholder: 'Enter your email',
          size,
          onInput: vi.fn(),
          onBlur: vi.fn(),
          onChange: vi.fn(),
        },
        slots: {
          icon: '<svg>mock icon</svg>',
        },
      })

      // Assert that the component is rendered
      expect(wrapper.exists()).toBe(true)

      // Assert that the icon slot is rendered
      expect(wrapper.find('svg').exists()).toBe(true)
      expect(wrapper.find('span.text-koromiko-600').exists()).toBe(true)

      // Assert snapshot
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
