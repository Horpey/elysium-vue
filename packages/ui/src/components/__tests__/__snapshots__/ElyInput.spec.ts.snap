// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`elyInput > icon > size LARGE 1`] = `
"<span class=\\"relative\\"><input class=\\"border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900 text-base px-4 py-3 pl-10\\" name=\\"email\\" type=\\"email\\" placeholder=\\"Enter your email\\"><span class=\\"pointer-events-none absolute inset-y-0 start-0 flex h-full items-center text-koromiko-600 w-10 pl-3 [&amp;>*]:h-6 [&amp;>svg]:w-6\\"><svg>mock icon</svg></span>
<!--v-if--></span>"
`;

exports[`elyInput > icon > size MEDIUM 1`] = `
"<span class=\\"relative\\"><input class=\\"border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900 text-sm px-3 py-2 pl-8\\" name=\\"email\\" type=\\"email\\" placeholder=\\"Enter your email\\"><span class=\\"pointer-events-none absolute inset-y-0 start-0 flex h-full items-center text-koromiko-600 w-8 pl-2 [&amp;>*]:h-4 [&amp;>svg]:w-4\\"><svg>mock icon</svg></span>
<!--v-if--></span>"
`;

exports[`elyInput > icon > size SMALL 1`] = `
"<span class=\\"relative\\"><input class=\\"border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900 text-xs px-2 py-1 pl-6\\" name=\\"email\\" type=\\"email\\" placeholder=\\"Enter your email\\"><span class=\\"pointer-events-none absolute inset-y-0 start-0 flex h-full items-center text-koromiko-600 w-6 pl-1 [&amp;>*]:h-2 [&amp;>svg]:w-2\\"><svg>mock icon</svg></span>
<!--v-if--></span>"
`;

exports[`elyInput > renders with custom props and triggers input event 1`] = `"<span class=\\"relative\\"><input class=\\"border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900 custom-input-class text-sm px-3 py-2\\" name=\\"email\\" type=\\"email\\" required=\\"\\" placeholder=\\"Enter your email\\"><!--v-if--><!--v-if--></span>"`;

exports[`elyInput > renders with default props 1`] = `"<span class=\\"relative\\"><input class=\\"border border-koromiko-600 shadow-sm focus:ring-1 focus:ring-koromiko-700 focus:border-koromiko-700 outline-none transition-all duration-200 ease-in-out w-full disabled:cursor-not-allowed disabled:bg-koromiko-800/10 disabled:text-koromiko-900 text-sm px-3 py-2\\" type=\\"text\\"><!--v-if--><!--v-if--></span>"`;
